Implement `system.server_settings`

pufit opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `system.server_settings` (similar to `system.settings`), which will contain server configurations.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)


https://github.com/ClickHouse/ClickHouse/issues/46231

pufit committed:
Implement `system.server_settings`

pufit committed:
Additional tests

serxa reviewed (commented):
In config reload we changed the code in several places from `has("XXX")` to `XXX.changed`. Which is a little bit different semantics. So if any of these settings are used to initialize stuff for the first time, it would not work. This is a little bit dangerous. But as far as I checked it is not used for initialization, so should be fine.

pufit commented:
> In config reload we changed the code in several places from `has("XXX")` to `XXX.changed`. Which is a little bit different semantics. So if any of these settings are used to initialize stuff for the first time, it would not work. This is a little bit dangerous. But as far as I checked it is not used for initialization, so should be fine.

As far as I understand, the behavior should be the same since we are loading ServerSettings from the configuration on each reload. So if some config is present in the file, it will have a `changed` attribute in ServerSettings, so the semantics should be the same.

pufit committed:
Update programs/server/Server.cpp

Co-authored-by: Sergei Trifonov <sergei@clickhouse.com>

pufit committed:
Update src/Core/ServerSettings.h

Co-authored-by: Sergei Trifonov <sergei@clickhouse.com>

serxa commented:
> As far as I understand, the behavior should be the same since we are loading ServerSettings from the configuration on each reload. So if some config is present in the file, it will have a changed attribute in ServerSettings, so the semantics should be the same.
If this is the case, then "changed" column has meaning "we have reloaded it from config and it was there". Basically most column will be marked as changed. I'm not sure what meaning do we actually need, but it looks counter intuitive at the moment.

pufit committed:
Removed unnecessary ifs

pufit commented:
> > As far as I understand, the behavior should be the same since we are loading ServerSettings from the configuration on each reload. So if some config is present in the file, it will have a changed attribute in ServerSettings, so the semantics should be the same.
> 
> If this is the case, then "changed" column has meaning "we have reloaded it from config and it was there". So basically most column will be marked as changed. I'm not sure what meaning do we actually need, but it looks counter intuitive at the moment.

I think you are right. And since most of the columns will be marked as changed, I guess we can simply remove those ifs (since all values will be present in server_settings)

serxa commented:
There are two small, but related problems:
 - we need docs for all system tables;
 - in docs we have to somehow explain what "changed" column mean.

For usual setting "changed" column was pretty obvious. If you changed a value using `SET` it is changed. I think it may be good to avoid setting "changed" to true during load if it has not been changed actually.

pufit commented:
> 
> For usual setting "changed" column was pretty obvious. If you changed a value using `SET` it is changed. I think it may be good to avoid setting "changed" to true during load if it has not been changed actually.

But if you specify some settings in `users.xml`, the value will also be `changed`, even if `SET` was never called. So in my understanding, `changed` means that the value was specified (in the config, for example). And also, `SET` can't change server settings, so we should either rename this column or completely remove it.

serxa commented:
> But if you specify some settings in users.xml, the value will also be changed, even if SET was never called.
Hm.. you mean a value in user's profile? Interesting, but it also looks confusing to me :)
> we should either rename this column or completely remove it
I'd suggest to remove it for now, can be added later with some clear and useful semantics.

serxa commented:
But if we remove it just from this system stable, it will also be not very consistent with other tables which contain settings. If we rename it, we should rename it in other tables for consistency. Both options are not very good. Okay, probably the best thing for now is just to leave it as is, as "changed" column meaning "value is specified".

serxa reviewed (approved):
Could you also add docs for this feature?

pufit committed:
Fix description typo

pufit committed:
en documentation

pufit committed:
ru documentation

rschu1ze commented on src/Storages/System/StorageSystemServerSettings.cpp:
```diff
@@ -0,0 +1,37 @@
+{
+    return {
+        {"name", std::make_shared<DataTypeString>()},
+        {"value", std::make_shared<DataTypeString>()},
```
Not sure but if the default value is available, it would be nice to include it. (same for system table `settings`)

pufit replied:
`system.settings` doesn't have a column with default values, but I can probably add it too.

pufit committed:
Add `default` column for `system.settings` and `system.server_settings`

rschu1ze commented on docs/en/operations/system-tables/server_settings.md:
````diff
@@ -0,0 +1,52 @@
+
+``` text
+┌─name─────────────────────────┬─value─┬─default─┬─changed─┬─description───────────────────────────┬─type───┐
+│ max_thread_pool_size         │ 10000 │ 10000   │       0 │ Max size for global thread pool.      │ UInt64 │
````
Guess that for illustration purposes, the table should also include changed settings.

pufit committed:
Update 01221_system_settings.reference

pufit committed:
Update server_settings.md

pufit committed:
Update server_settings.md

pufit committed:
fix test

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+#define SERVER_SETTINGS(M, ALIAS) \
+    M(Bool, show_addresses_in_stack_traces, true, "If it is set true will show addresses in stack traces", 0) \
+    M(Bool, shutdown_wait_unfinished_queries, false, "If set true ClickHouse will wait for running queries finish before shutdown.", 0) \
+    M(UInt64, max_thread_pool_size, 10000, "Max size for global thread pool.", 0) \
```
```suggestion
    M(UInt64, max_thread_pool_size, 10000, "The maximum number of threads that could be allocated from the OS and used for query execution and background operations.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(Bool, show_addresses_in_stack_traces, true, "If it is set true will show addresses in stack traces", 0) \
+    M(Bool, shutdown_wait_unfinished_queries, false, "If set true ClickHouse will wait for running queries finish before shutdown.", 0) \
+    M(UInt64, max_thread_pool_size, 10000, "Max size for global thread pool.", 0) \
+    M(UInt64, max_thread_pool_free_size, 1000, "Max free size for global thread pool.", 0) \
```
```suggestion
    M(UInt64, max_thread_pool_free_size, 1000, "The maximum number of threads that will always stay in a global thread pool once allocated.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(Bool, shutdown_wait_unfinished_queries, false, "If set true ClickHouse will wait for running queries finish before shutdown.", 0) \
+    M(UInt64, max_thread_pool_size, 10000, "Max size for global thread pool.", 0) \
+    M(UInt64, max_thread_pool_free_size, 1000, "Max free size for global thread pool.", 0) \
+    M(UInt64, thread_pool_queue_size, 10000, "Queue size for global thread pool.", 0) \
```
```suggestion
    M(UInt64, thread_pool_queue_size, 10000, "The maximum number of tasks that will be placed in a queue and wait for execution.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, max_thread_pool_size, 10000, "Max size for global thread pool.", 0) \
+    M(UInt64, max_thread_pool_free_size, 1000, "Max free size for global thread pool.", 0) \
+    M(UInt64, thread_pool_queue_size, 10000, "Queue size for global thread pool.", 0) \
+    M(UInt64, max_io_thread_pool_size, 100, "Max size for IO thread pool.", 0) \
```
```suggestion
    M(UInt64, max_io_thread_pool_size, 100, "The maximum number of threads that would be used for IO operations", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(Int32, max_connections, 1024, "Max server connections.", 0) \
+    M(UInt32, asynchronous_metrics_update_period_s, 1, "Period in seconds for updating asynchronous metrics.", 0) \
+    M(UInt32, asynchronous_heavy_metrics_update_period_s, 120, "Period in seconds for updating asynchronous metrics.", 0) \
+    M(UInt32, max_threads_for_connection_collector, 100, "Max thread count for connection collector.", 0) \
```
The default value was 10

nikitamikhaylov replied:
```suggestion
    M(UInt32, max_threads_for_connection_collector, 10, "The maximum number of thread that will be used for draining connections asynchronously in a background upon finishing executing distributed queries.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt32, max_threads_for_connection_collector, 100, "Max thread count for connection collector.", 0) \
+    M(String, default_database, "default", "Default database name.", 0) \
+    M(String, tmp_policy, "", "Policy for storage with temporary data.", 0) \
+    M(UInt64, max_temporary_data_on_disk_size, 0, "Max data size for temporary storage.", 0) \
```
```suggestion
    M(UInt64, max_temporary_data_on_disk_size, 0, "The maximum amount of storage that could be used for external aggregation, joins or sorting., ", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, concurrent_threads_soft_limit_num, 0, "Sets how many concurrent thread can be allocated before applying CPU pressure. Zero means Unlimited.", 0) \
+    M(UInt64, concurrent_threads_soft_limit_ratio_to_cores, 0, "Same as concurrent_threads_soft_limit_num, but with ratio to cores.", 0) \
+    \
+    M(UInt64, background_pool_size, 16, "Sets background pool size.", 0) \
```
```suggestion
    M(UInt64, background_pool_size, 16, "The maximum number of threads what will be used for merging or mutating data part for MergeTree tables in a background", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, concurrent_threads_soft_limit_ratio_to_cores, 0, "Same as concurrent_threads_soft_limit_num, but with ratio to cores.", 0) \
+    \
+    M(UInt64, background_pool_size, 16, "Sets background pool size.", 0) \
+    M(UInt64, background_merges_mutations_concurrency_ratio, 2, "Sets background merges mutations concurrency ratio.", 0) \
```
```suggestion
    M(UInt64, background_merges_mutations_concurrency_ratio, 2, "The multiplier which shows the relation between the number of tasks that could be executed concurrently and the number of threads being used.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    \
+    M(UInt64, background_pool_size, 16, "Sets background pool size.", 0) \
+    M(UInt64, background_merges_mutations_concurrency_ratio, 2, "Sets background merges mutations concurrency ratio.", 0) \
+    M(String, background_merges_mutations_scheduling_policy, "round_robin", "Sets background merges mutations scheduling policy.", 0) \
```
```suggestion
    M(String, background_merges_mutations_scheduling_policy, "round_robin", "The policy on how to perform a scheduling for background merges and mutations. Possible values are: `round_robin` and `shortest_task_first`. ", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_pool_size, 16, "Sets background pool size.", 0) \
+    M(UInt64, background_merges_mutations_concurrency_ratio, 2, "Sets background merges mutations concurrency ratio.", 0) \
+    M(String, background_merges_mutations_scheduling_policy, "round_robin", "Sets background merges mutations scheduling policy.", 0) \
+    M(UInt64, background_move_pool_size, 8, "Sets background move pool size.", 0) \
```
```suggestion
    M(UInt64, background_move_pool_size, 8, "The maximum number of threads that will be used for performing moving data parts to another disk or volume for MergeTree tables in a background.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_merges_mutations_concurrency_ratio, 2, "Sets background merges mutations concurrency ratio.", 0) \
+    M(String, background_merges_mutations_scheduling_policy, "round_robin", "Sets background merges mutations scheduling policy.", 0) \
+    M(UInt64, background_move_pool_size, 8, "Sets background move pool size.", 0) \
+    M(UInt64, background_fetches_pool_size, 8, "Sets background fetches pool size.", 0) \
```
```suggestion
    M(UInt64, background_fetches_pool_size, 8, "The maximum number of threads that will be used for fetching data parts from another replica for MergeTree tables in a background.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(String, background_merges_mutations_scheduling_policy, "round_robin", "Sets background merges mutations scheduling policy.", 0) \
+    M(UInt64, background_move_pool_size, 8, "Sets background move pool size.", 0) \
+    M(UInt64, background_fetches_pool_size, 8, "Sets background fetches pool size.", 0) \
+    M(UInt64, background_common_pool_size, 8, "Sets background common pool size.", 0) \
```
```suggestion
    M(UInt64, background_common_pool_size, 8, "The maximum number of threads that will be used for performing a variety of operations (mostly garbage collection) for MergeTree tables in a background.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_move_pool_size, 8, "Sets background move pool size.", 0) \
+    M(UInt64, background_fetches_pool_size, 8, "Sets background fetches pool size.", 0) \
+    M(UInt64, background_common_pool_size, 8, "Sets background common pool size.", 0) \
+    M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Sets background flush schedule pool size.", 0) \
```
```suggestion
    M(UInt64, background_buffer_flush_schedule_pool_size, 16, "The maximum number of threads that will be used for performing flush operations for Buffer-engine tables in a background.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_fetches_pool_size, 8, "Sets background fetches pool size.", 0) \
+    M(UInt64, background_common_pool_size, 8, "Sets background common pool size.", 0) \
+    M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Sets background flush schedule pool size.", 0) \
+    M(UInt64, background_schedule_pool_size, 16, "Sets background schedule pool size.", 0) \
```
```suggestion
    M(UInt64, background_schedule_pool_size, 16, "The maximum number of threads that will be used for constantly executing some small periodic operations.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_common_pool_size, 8, "Sets background common pool size.", 0) \
+    M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Sets background flush schedule pool size.", 0) \
+    M(UInt64, background_schedule_pool_size, 16, "Sets background schedule pool size.", 0) \
+    M(UInt64, background_message_broker_schedule_pool_size, 16, "Sets background message broker schedule pool size.", 0) \
```
```suggestion
    M(UInt64, background_message_broker_schedule_pool_size, 16, "The maximum number of threads that will be used for executing background operations for message streaming.", 0) \
```

nikitamikhaylov commented on src/Core/ServerSettings.h:
```diff
@@ -0,0 +1,78 @@
+    M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Sets background flush schedule pool size.", 0) \
+    M(UInt64, background_schedule_pool_size, 16, "Sets background schedule pool size.", 0) \
+    M(UInt64, background_message_broker_schedule_pool_size, 16, "Sets background message broker schedule pool size.", 0) \
+    M(UInt64, background_distributed_schedule_pool_size, 16, "Sets background distributed schedule pool size.", 0) \
```
```suggestion
    M(UInt64, background_distributed_schedule_pool_size, 16, "The maximum number of threads that will be used for executing distributed sends.", 0) \
```

pufit committed:
Apply suggestions from code review

Co-authored-by: Nikita Mikhaylov <mikhaylovnikitka@gmail.com>

nikitamikhaylov commented on docs/en/operations/system-tables/server_settings.md:
````diff
@@ -0,0 +1,52 @@
+```
+
+``` text
+┌─name─────────────────────────┬─value─┬─default─┬─changed─┬─description───────────────────────────┬─type───┐
````
Need to replace this output, because description was changed in the code.

pufit committed:
Update docs

tavplubix commented:
@pufit, please check failed CI checks. Even if failures are unrelated, it still makes sense to check and create issues (if the corresponding issues don't exist yet). For example, Stress test (ubsan) found a crash, it has to be investigated.

pufit commented:
> @pufit, please check failed CI checks. Even if failures are unrelated, it still makes sense to check and create issues (if the corresponding issues don't exist yet). For example, Stress test (ubsan) found a crash, it has to be investigated.

:thinking_face: I thought it was some sort of flaky test because I didn't see them in the initial commit, and since that, nothing major was changed. I'll try to take a look at it and will create an issue if it isn't related to my PR (or rollback if it is related)

Thank you for noticing it!

tavplubix commented:
> I thought it was some sort of flaky test

Yes, unfortunately, we have a lot of flaky functional and integration tests lately (btw, there's a good [query](https://play.clickhouse.com/play?user=play#d2l0aCAKJyUwMjExNl9pbnRlcmFjdGl2ZV9oZWxsbyUnIGFzIG5hbWVfcGF0dGVybiwKJyUlJyBhcyBjb250ZXh0X3BhdHRlcm4sCicyMDIyLTA5LTAxJyBhcyBzdGFydF9kYXRlLAooNDU0NjEpIGFzIG5vaXN5X3BycywKKCdTdGF0ZWxlc3MgdGVzdHMgKGFzYW4pJywgJ1N0YXRlbGVzcyB0ZXN0cyAoYWRkcmVzcyknLCAnU3RhdGVsZXNzIHRlc3RzIChhZGRyZXNzLCBhY3Rpb25zKScpIGFzIGJhY2twb3J0X2FuZF9yZWxlYXNlX3NwZWNpZmljX2NoZWNrcwpzZWxlY3QgCnRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBhcyBkLApjb3VudCgpLCAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlciksIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSBzdGFydF9kYXRlIDw9IGNoZWNrX3N0YXJ0X3RpbWUgYW5kIHB1bGxfcmVxdWVzdF9udW1iZXIgbm90IGluIAooc2VsZWN0IHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHJuIGZyb20gY2hlY2tzIHdoZXJlIHBybiE9MCBhbmQgc3RhcnRfZGF0ZSA8PSBjaGVja19zdGFydF90aW1lIGFuZCBjaGVja19uYW1lIGluIGJhY2twb3J0X2FuZF9yZWxlYXNlX3NwZWNpZmljX2NoZWNrcykgCmFuZCB0ZXN0X25hbWUgbGlrZSBuYW1lX3BhdHRlcm4gYW5kIHRlc3RfY29udGV4dF9yYXcgaWxpa2UgY29udGV4dF9wYXR0ZXJuCmFuZCBwdWxsX3JlcXVlc3RfbnVtYmVyIG5vdCBpbiBub2lzeV9wcnMKYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScsICdmYWlsdXJlJykgZ3JvdXAgYnkgZCBvcmRlciBieSBkIGRlc2M=) to check if a test if flaky or not and also there's https://aretestsgreenyet.com/). But anyway, it makes sense to take a look at flaky tests and try to figure out why these tests are flaky/where these tests were broken. If it's hard to figure out at the first glance - ask test's author to take a look. 

As for Stress tests and Fuzzers, they are highly randomized and usually find unrelated but major issues (like crashes and deadlocks). These failures always require attention.
